presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -46,4 +48,34 @@ default boolean isEnabled(Session session)
 
         Session getSession();
     }
+
+    class Result
+    {
+        public static Result empty()
+        {
+            return new Result(Optional.empty());
+        }
+
+        public static Result addPlanNode(PlanNode transformedPlan)
```

Comment:
Why "add"? This should probably be "from(...)" or "of(...)"

Reply:
Because: https://github.com/prestodb/presto/pull/8851#discussion_r135466510

It used to be `replace`, was not that better than `of`? IMO `of` does not tell much.

Reply:
The way I think about it is that the rule constructs the result (which is just a holder). The act of adding or replacing is something the optimizer/memo do as part of *their* logic.